### ISSUE TITLE
Must the following pseudocode be supported by all JAX-RS implementations?

### ISSUE THREAD

**Issue by ljnelson** [eclipse-ee4j/jaxrs-api#698]
I'm exploring some edge cases that come up as a result of CDI reaching version 2.0 and now containing the ability to start a CDI container in Java SE.

I'm curious if the following pseudocode case MUST be supported by all JAX-RS implementations.

(I'm typing this in the issue editor so please forgive syntax errors etc.  I believe the intent of my question will be clear nonetheless.)

A root resource class that I intend to be compliant with section 3.1:
```
@Path("/hello")
@RequestScoped // from CDI; a bean-defining annotation
public class Greet {
  @GET
  @Produces("text/plain")
  public String greet() {
    return "Hello!";
  }
}
```
I start my Java program (Java SE, not Java EE or Jakarta EE, no Servlet, etc.) by doing:
```
SeContainerInitializer.newInstance().disableDiscovery().addBeanClasses(Greet.class, MyStarter.class).initialize();
```
(The `disableDiscovery()` call is just to take CDI's discovery mechanism out of the discussion.  You'll note the `Greet.class` CDI bean is programmatically added here.)

Inside the `MyStarter.class` (also programmatically added):
```
@ApplicationScoped
final class MyStarter {
  private final void onStartup(@Observes @Initialized(ApplicationScoped.class) final Object startupEvent) {
    // Assume myJaxRsImplementationWithCDIIntegrationSupport is a
    // JAX-RS 2.1.1 implementation that has support for CDI in some
    // fashion, and this call starts it up.
    myJaxRsImplementationWithCDIIntegrationSupport.start();
  }
}
```
So when the CDI container comes up, `myJaxRsImplementationWithCDIIntegrationSupport` is started.

My question is:

**MUST `Greet` be a root resource class in the resulting JAX-RS application?**

Section 11.2.3 reads in part: "In a product that supports CDI [it is my intention that the above pseudocode constitutes such a situation], implementations MUST support the use of CDI-style Beans as root resource classes, providers and Application subclasses."

I am trying to understand specifically whether a root resource class present as a CDI bean (in this case [ignoring bean discovery](http://docs.jboss.org/cdi/api/2.0/javax/enterprise/inject/se/SeContainerInitializer.html#disableDiscovery--)) must be used even in the absence of an `Application` subclass or when an `Application` returns `null` or empty values from its `getClasses()` and `getSingletons()` methods.

This is tangentially related to #409.

**Comment by mkarg** [comment]
I think there is a misunderstanding:
* "CDI" (in the sense of 11.2.3) means "CDI 1.0 Application-Server-Provided Container", but your code uses "CDI 2.0 Java SE Container" which is so far not clearly specified by the JAX-RS spec (we're working on that one).
* "product" (in the sense of 11.2.3) is a framework of application server like Jersey, RESTeasy or CXF, not your custom code.
* That spec chaper relates to the action starting a "product" (i. e. booting Jersey), not starting your CDI custom code.
* What you strive for (startup of custom code) is supported with JAX-RS 2.2, which defines a Java SE Bootstrap API, but which does not support CDI 2.0 so far (will work in Jersey 2.29 or 2.30, but not 2.27 or 2.28).

So no, JAX-RS 2.1 and 2.2 currently will not support that code, but a *future release* of the JAX-RS specification will demand cross-product support.

**Comment by ljnelson** [comment]
I understand you, but I don't think the CDI version matters.

Let me rephrase the question:

I have a build of (say) Glassfish that supports JAX-RS 2.0.  I add a jar file to its classpath using `asadmin`.  The jar file contains nothing other than `Greet.class` above.  I start Glassfish.  Must my root resource class be recognized and part of a JAX-RS application?  Must this be true of all JAX-RS implementations embedded in such application servers?

If the answer is no, as I suspect, why not?

**Comment by mkarg** [comment]
The CDI version matters, because CDI 2.0 knows Java SE containers, while CDI 1.0 does not. Your code looks like using the JAX-RS 2.2's Java SE Bootstrap, which is different from JAX-RS 2.1's Server Container bootstrap.

The answer is *no*. Products COULD do that but they are not forced to, at least not by the *JAX-RS* specification.

The JAX-RS specification does not cover the case of adding JAX-RS resources *to the product's classpath* (the PDF not even contains the word "classpath"). It only covers adding JAX-RS resources to *containers* powered by products. Typically these products, in particular application servers, have a rather complex hierarchy of classloaders, where the inside of containers is more or less shielded from the underlying product. So it could happen that some products present that class to the JAX-RS container, while others do not. See, adding classes *to applications* is different from adding classes *to products*.
